src/java.desktop/share/classes/javax/swing/JMenu.java

```diff
@@ -478,6 +478,16 @@ protected Point getPopupMenuOrigin() {
                 y = 0 - yOffset - pmSize.height;   // Otherwise drop 'up'
             }
         }
+        // Note that the y position may be later adjusted to fit the menu into the screen if possible.
+        // However, the code that does it (JPopupMenu.adjustPopupLocationToFitScreen) has no idea which screen
+        // to fit into, and determines it by the position calculated here, so we need to make sure it's on
+        // the correct screen, otherwise the menu may appear on the wrong screen (JDK-6415065).
+        if (position.y + y < screenBounds.y) { // Above the current screen?
+            y = screenBounds.y - position.y; // Fit into the screen, relative to our origin.
+        }
+        if (position.y + y >= screenBounds.y + screenBounds.height) { // Below the current screen?
```

Comment:
why only the y coordinate is checked? Can we implement it for x as well for symmetry?

Reply:
A good idea. I was actually thinking about it after I had submitted this. Even though I've never encountered any issues with the X coordinate, I think it's still possible in theory in some unusual circumstances. Anyway, an extra check definitely won't hurt. Will do.

Reply:
Added the X coordinate checks as well. While I was testing it, I noticed that sometimes the screen bounds are already wrong here because the graphics configuration is returned for the wrong display above. It happens, for example, in Ubuntu, when a window spans across two monitors. This seems to be a separate issue, though, and I can't do anything about it here, it's something deeper inside the internals, and likely is platform-dependent.